### PR TITLE
Adds membership and ownership fields for Dynamic Access List

### PR DESCRIPTION
**Pull request by tcsc**
In preparation for dynamic access lists, this PR adds fields for discriminating
dynamic lists from the original, "static" access lists.

### CODE REVIEW

**Review by mdwn — APPROVED** [review]
Great work!

File: api/types/accesslist/accesslist.go [hunk 1]
```diff
@@ -42,6 +42,13 @@ const (
 	twoWeeks = 24 * time.Hour * 14
 )
 
+type Inclusion string
+
+const (
+	Implicit Inclusion = "implicit"
+	Explicit Inclusion = "explicit"
```
Comment (r0mant) [inline api/types/accesslist/accesslist.go:49]: Nit: godocs

File: api/types/accesslist/accesslist.go [hunk 2]
```diff
@@ -241,6 +266,18 @@ func NewAccessList(metadata header.Metadata, spec Spec) (*AccessList, error) {
 	return accessList, nil
 }
 
+func checkInclusion(i Inclusion) (Inclusion, error) {
+	if i == "" {
+		return Explicit, nil
+	}
+
+	if i != Explicit && i != Implicit {
+		return Explicit, trace.BadParameter("invalid inclusion mode %q", i)
```
Comment (r0mant) [inline api/types/accesslist/accesslist.go:275]: Nit: Expand the message to mention which ones are valid?

File: api/types/accesslist/accesslist.go [hunk 3]
```diff
@@ -241,6 +266,18 @@ func NewAccessList(metadata header.Metadata, spec Spec) (*AccessList, error) {
 	return accessList, nil
 }
 
+func checkInclusion(i Inclusion) (Inclusion, error) {
+	if i == "" {
+		return Explicit, nil
+	}
+
+	if i != Explicit && i != Implicit {
```
Comment (r0mant) [inline api/types/accesslist/accesslist.go:274]: Nit: I think `switch` statement would work really well here instead.

File: api/proto/teleport/accesslist/v1/accesslist.proto [hunk 4]
```diff
@@ -63,6 +63,24 @@ message AccessListSpec {
 
   // title is a plaintext short description of the access list.
   string title = 8;
+
+  // membership defines how list membership is applied. There are two possible
+  // values:
+  //   `explicit` (default): To be considered ag member of the access list, a user
+  //                         must both meet the `membership_conditions` AND be
+  //                         explicitly added to the list.
+  //   implicit: Any user meeting the `membership_conditions` will automatically
```
Comment (r0mant) [inline api/proto/teleport/accesslist/v1/accesslist.proto:72]:
```suggestion
  //   implicit: Any user meeting the `membership_requires` conditions will automatically
```

File: api/proto/teleport/accesslist/v1/accesslist.proto [hunk 5]
```diff
@@ -63,6 +63,24 @@ message AccessListSpec {
...
+  //             be cosidered a member of this list.
+  string membership = 9;
+
+  // ownership defines how list ownership of this list is determined. There are
+  // two possible values:
+  //   `explicit` (default): To be considered an owner of the access list, a user
+  //                         must both meet the `ownership_conditions` AND be
+  //                         explicitly added to the list.
+  //   implicit: Any user meeting the `ownership_conditions` will automatically
```
Comment (r0mant) [inline api/proto/teleport/accesslist/v1/accesslist.proto:81]:
```suggestion
  //   implicit: Any user meeting the `ownership_requires` conditions will automatically
```

File: api/proto/teleport/accesslist/v1/accesslist.proto [hunk 6]
```diff
@@ -63,6 +63,24 @@ message AccessListSpec {
...
+  //                         must both meet the `membership_conditions` AND be
+  //                         explicitly added to the list.
+  //   implicit: Any user meeting the `membership_conditions` will automatically
+  //             be cosidered a member of this list.
+  string membership = 9;
+
+  // ownership defines how list ownership of this list is determined. There are
+  // two possible values:
+  //   `explicit` (default): To be considered an owner of the access list, a user
+  //                         must both meet the `ownership_conditions` AND be
```
Comment (r0mant) [inline api/proto/teleport/accesslist/v1/accesslist.proto:79]:
```suggestion
  //                         must both meet the `ownership_requires` conditions AND be
```

File: api/proto/teleport/accesslist/v1/accesslist.proto [hunk 7]
```diff
@@ -63,6 +63,24 @@ message AccessListSpec {
 
   // title is a plaintext short description of the access list.
   string title = 8;
+
+  // membership defines how list membership is applied. There are two possible
+  // values:
+  //   `explicit` (default): To be considered ag member of the access list, a user
+  //                         must both meet the `membership_conditions` AND be
```
Comment (r0mant) [inline api/proto/teleport/accesslist/v1/accesslist.proto:70]:
```suggestion
  //                         must both meet the `membership_requires` conditions AND be
```

**Review by r0mant — APPROVED** [review]

**Pull request merged**